File: lib/services/inventario_api.dart
import 'package:lh_tonner/services/api_client.dart';

/// Modelo de ítem de inventario (tabla INVENTARIO_LH_TONER + datos del producto).
class InventarioItem {
  const InventarioItem({
    required this.idInventario,
    required this.cantidad,
    required this.fkIdProductos,
    this.nombre,
    this.productoNombre,
    this.productoCategoria,
  });

  final int idInventario;
  final int cantidad;
  final int fkIdProductos;
  final String? nombre;
  final String? productoNombre;
  final String? productoCategoria;

  /// Nombre para mostrar en lista (producto_nombre del JOIN o nombre del ítem).
  String get displayNombre => productoNombre ?? nombre ?? 'Sin nombre';

  factory InventarioItem.fromJson(Map<String, dynamic> json) {
    return InventarioItem(
      idInventario: (json['id_inventario'] ?? json['id'] ?? 0) as int,
      cantidad: (json['cantidad'] ?? 0) as int,
      fkIdProductos: (json['fk_ID_PRODUCTOS_LH_TONER'] ?? json['id_producto'] ?? json['fk_id_productos_lh_toner'] ?? 0) as int,
      nombre: json['nombre'] as String?,
      productoNombre: (json['producto_nombre'] ?? json['productoNombre']) as String?,
      productoCategoria: (json['producto_categoria'] ?? json['productoCategoria']) as String?,
    );
  }
}

/// API para INVENTARIO_LH_TONER (listar, añadir stock, actualizar, eliminar).
class InventarioApi {
  InventarioApi._();

  static const String _path = '/api/inventario_lh_toner';

  /// Lista el inventario (incluye producto_nombre, producto_categoria para la vista).
  static Future<List<InventarioItem>> listar() async {
    final res = await ApiClient.get(_path);
    if (!res.ok || res.data == null) return [];
    List? list;
    if (res.data is List) {
      list = res.data as List;
    } else if (res.data is Map) {
      final m = res.data as Map;
      list = (m['data'] ?? m['inventario'] ?? m['items'] ?? m['results']) as List?;
    }
    if (list == null) return [];
    return list
        .map((e) => InventarioItem.fromJson(Map<String, dynamic>.from(e as Map)))
        .toList();
  }

  /// Anade stock. Body: fk_ID_PRODUCTOS_LH_TONER (o id_producto) + cantidad.
  static Future<({bool ok, String? message})> anadirStock(int idProducto, int cantidad) async {
    final res = await ApiClient.post(_path, {
      'fk_ID_PRODUCTOS_LH_TONER': idProducto,
      'id_producto': idProducto,
      'cantidad': cantidad,
    });
    if (res.ok) return (ok: true, message: null);
    return (ok: false, message: res.message ?? 'Error al anadir stock');
  }

  /// Actualiza un registro (cantidad y/o producto).
  static Future<({bool ok, String? message})> actualizar(
      int idInventario, {int? cantidad, int? fkIdProducto}) async {
    final body = <String, dynamic>{};
    if (cantidad != null) body['cantidad'] = cantidad;
    if (fkIdProducto != null) {
      body['fk_ID_PRODUCTOS_LH_TONER'] = fkIdProducto;
      body['id_producto'] = fkIdProducto;
    }
    final res = await ApiClient.put('$_path/$idInventario', body.isEmpty ? null : body);
    if (res.ok) return (ok: true, message: null);
    return (ok: false, message: res.message ?? 'Error al actualizar');
  }

  /// Elimina un registro de inventario.
  static Future<({bool ok, String? message})> eliminar(int idInventario) async {
    final res = await ApiClient.delete('$_path/$idInventario');
    if (res.ok) return (ok: true, message: null);
    return (ok: false, message: res.message ?? 'Error al eliminar');
  }
}

File: src/services/inventarioApi.js
import http from './apiClient'

// ── Stock actual ────────────────────────────────────────────────────────────
export async function listarStock() {
  try {
    const res = await http.get('/api/inventario/stock')
    const raw = Array.isArray(res.data) ? res.data : (res.data?.data || res.data?.items || [])
    return raw.map((r) => ({
      idProducto:      r.id_producto ?? r.idProducto ?? 0,
      stockActual:     Number(r.stock_actual ?? r.stockActual ?? 0),
      nombreProducto:  r.nombre_producto ?? r.nombreProducto ?? r.producto ?? '',
      nombreCategoria: r.nombre_categoria ?? r.nombreCategoria ?? r.categoria ?? '',
    }))
  } catch (err) {
    return { error: err.response?.data?.message || 'Error al cargar stock' }
  }
}

// ── Categorías ──────────────────────────────────────────────────────────────
export async function listarCategorias() {
  try {
    const res = await http.get('/api/inventario/categorias')
    const raw = Array.isArray(res.data) ? res.data : (res.data?.data || [])
    return raw.map((c) => ({
      id:     c.id_categoria ?? c.id ?? 0,
      nombre: c.nombre_categoria ?? c.nombre ?? '',
    }))
  } catch (_) { return [] }
}

// ── Productos ───────────────────────────────────────────────────────────────
export async function listarProductos() {
  try {
    const res = await http.get('/api/inventario/productos')
    const raw = Array.isArray(res.data) ? res.data : (res.data?.data || [])
    return raw.map((p) => ({
      idProducto:      p.id_producto ?? p.id ?? 0,
      nombreProducto:  p.nombre_producto ?? p.nombre ?? '',
      nombreCategoria: p.nombre_categoria ?? p.nombreCategoria ?? '',
      idCategoria:     p.id_categoria ?? p.idCategoria ?? null,
    }))
  } catch (_) { return [] }
}

export async function crearProducto(nombre, idCategoria) {
  try {
    await http.post('/api/inventario/productos', { nombre_producto: nombre, id_categoria: idCategoria })
    return { ok: true }
  } catch (err) {
    return { ok: false, message: err.response?.data?.message || 'Error al crear producto' }
  }
}

export async function actualizarProducto(id, nombre, idCategoria) {
  try {
    await http.put(`/api/inventario/productos/${id}`, { nombre_producto: nombre, id_categoria: idCategoria })
    return { ok: true }
  } catch (err) {
    return { ok: false, message: err.response?.data?.message || 'Error al actualizar' }
  }
}

export async function eliminarProducto(id) {
  try {
    await http.delete(`/api/inventario/productos/${id}`)
    return { ok: true }
  } catch (err) {
    return { ok: false, message: err.response?.data?.message || 'Error al eliminar' }
  }
}

export async function ajustarStock(idProducto, nuevoStock) {
  try {
    const res = await http.post(`/api/inventario/stock/${idProducto}/ajuste`, { nuevo_stock: nuevoStock })
    return { ok: true, data: res.data }
  } catch (err) {
    return { ok: false, message: err.response?.data?.error || err.response?.data?.message || 'Error al ajustar stock' }
  }
}
